components/patient-data/ui/src/main/resources/PhenoTips/JSONImportDialog.xml

```diff
@@ -0,0 +1,519 @@
+          var notificationMsg = "$escapetool.javascript($services.localization.render('phenotips.patientData.jsonImportDialog.importSuccess'))";
+          _this.$body._x_notification.replace(new XWiki.widgets.Notification(notificationMsg, "done"));
+          window.location.reload();
+        },
+        onFailure: function(response) {
+          var failureMsg = _this._generateFailureReason(response.status);
+          _this.$body._x_notification.replace(new XWiki.widgets.Notification(failureMsg, "error"));
+        }
+      });
+    },
+
+    _initFilePicker : function() {
+      if (!!window.FileReader &amp;&amp; !!window.FileList) {
+        // Show upload button if the browser supports FileReader/DOM File API
+        var uploadFileSelector = new Element('input', {"type" : "file", "id": 'jsonInputFile', "style": 'display:none'});
+        var _this = this;
+
+        uploadFileSelector.observe('change', function(event) {
+          _this.handleFileUpload(this.files);
+          try {
+            this.value = "";  // clear file selector
+          } catch (err) {
+            // some older browsers do not allow setting value of a file input element and may generate a security error
+          }
+        });
+
+        var fileUploadWrapper = new Element('div', {'class' : 'file-select-block-wrapper'});
+
+        var uploadButtonName = "$escapetool.javascript($services.localization.render('phenotips.patientData.buttons.importFileSelect'))";
+
+        var uploadPromptString = "$escapetool.javascript($services.localization.render('phenotips.patientData.jsonImportDialog.sectionTitles.import'))";
+
+        var uploadButton = '&lt;input class="button file-select-button" type="button" value="' + uploadButtonName + '"&gt;&lt;/input&gt;'
+
+        var re = /___placeholder___/gi;
+        uploadPromptString = uploadPromptString.replace(re, uploadButton);
+
+        fileUploadWrapper.insert(new Element('label', {'class': 'import-label'})
+          .update(uploadPromptString));
+
+        fileUploadWrapper.down("input.file-select-button").observe('click', function() {
+          var fileElem = document.getElementById("jsonInputFile");
+          fileElem.click();
+        });
+
+        this.mainDiv.insertBefore(fileUploadWrapper, this.jsonData).insert(uploadFileSelector);
+      } else {
+        this.mainDiv.insert(new Element('div', {'class': 'box infomessage import-warning-nofileupload'})
+          .update("$escapetool.javascript($services.localization.render('phenotips.patientData.jsonImportDialog.fileUploadNotSupported'))"));
+      }
+    },
+
+    _initServicePicker : function(jsonButton) {
+      for (var service in this.importServices) {
+        // Find the correct import service
+        if (jsonButton.hasClassName(service)) {
+          this._builder = this.importServices[service];
+        }
+      }
+    },
+
+    _generateFailureReason : function(status) {
+      var response;
+
+      switch (status) {
+        case 304 :
```

Comment:
As I said above, this isn't an error. You should treat it as a success, and for that you should say above in the `new Ajax.Request` configuration:

                   on304 : function(response) {
                     response.request.options.onSuccess(response);
                   },